This assignment requires writing a Makefile and a markdown report. 

Use the Makefile developed for your previous assignment, which contains targets for creating a BAM file.

Use the BAM file generated using reads from the SRA.

In your report, show the commands and the answers for each of these questions:

How many reads did not align with the reference genome?
How many primary, secondary, and supplementary alignments are in the BAM file?
How many properly-paired alignments on the reverse strand are formed by reads contained in the first pair ?
Make a new BAM file that contains only the properly paired primary alignments with a mapping quality of over 10
Compare the flagstats for your original and your filtered BAM file.
